### PR TITLE
Enhanced redis security: ACL and TLS

### PR DESCRIPTION
**Pull request by davidpil2002**
#### Why I did it
Enhance Redis Security.
Add Redis ACL support
Add Redis TLS support

| PR title | state | context |
| ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------ | 
| [[sonic-py-swsssdk] Added support for Redis](https://github.com/sonic-net/sonic-py-swsssdk/pull/144) | GitHub issue/pull request detail | GitHub pull request check contexts |
| [[sonic-swss-common] Added support for Redis](https://github.com/sonic-net/sonic-swss-common/pull/844) | GitHub issue/pull request detail | GitHub pull request check contexts |
| [[sonic-platform-common]Added support for Redis](https://github.com/sonic-net/sonic-platform-common/pull/424) | GitHub issue/pull request detail | GitHub pull request check contexts |
| [[redis-dump-load]Added support for Redis](https://github.com/p/redis-dump-load/pull/67) | GitHub issue/pull request detail | GitHub pull request check contexts |

##### Work item tracking
- Microsoft ADO **(number only)**:

#### How I did it
Add Redis ACL support by creating 2 users nad configure them in Redis Server.
one admin user who has access to any commands and operations like write/read
one monitor user who has access to read-only
Add Redis TLS by enable it in Redis Server.

#### How to verify it
Good flow:
each flow that exists today that uses the Redis DB should continue to be supported
Bad flow:
Add a new user which not belong to the admin group, and try to write to the Redis DB, this operation should be blocked.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

**Comment by davidpil2002** [comment]
The feature was built and tested locally when including all the submodule modifications.
Looks like all the dockers working correctly. But, required some adjustment in the sonic-mgmt test.
When we finish all the sonic-mgmt adjustments it will be possible to merge this feature.

For now, I would appreciate if you could review all the PRs related.
Thanks

**Comment by qiluo-msft** [comment]
In this PR's description, all the code PR link text is "Added support for LDAP". Is it a mistake?

File: files/scripts/gen-redis-certs.sh [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+[ server_cert ]
+keyUsage = digitalSignature, keyEncipherment
+nsCertType = server
+
+[ client_cert ]
+keyUsage = digitalSignature, keyEncipherment
+nsCertType = client
+_END_
+
+generate_cert redis "Generic-cert"
```
Comment (liuh-80) [inline files/scripts/gen-redis-certs.sh:52]: What's the cert file permission? as my understand the RW cert and CA cert can only be read by admin and RW user. the RO cert cab be read by anyone.

File: files/scripts/gen-redis-certs.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+cat > certs/tls/openssl.cnf <<_END_
+[ server_cert ]
+keyUsage = digitalSignature, keyEncipherment
+nsCertType = server
+
+[ client_cert ]
+keyUsage = digitalSignature, keyEncipherment
+nsCertType = client
+_END_
+
```
Comment (liuh-80) [inline files/scripts/gen-redis-certs.sh:51]: The CA cert need protect by file permission? for example only RW/Admin user can modify it?

**Comment by Yarden-Z** [comment]
@liuh-80 Any additional comments on this PR and the related PRs?

**Comment by liuh-80** [comment]
> @liuh-80 Any additional comments on this PR and the related PRs?

No for this PR, for related PRs I suggest:
1. Because sonic need user an SSL enabled hiredis_ssl, the build pipeline of other SONiC repo may failed because the hiredis_ssl debian package need install before build. you may need check and fix all pipeline before PR merge to prevent build break. I think most .yml file in this query may need update: https://github.com/search?q=org%3Asonic-net+libswsscommon_1.0.0_&type=code

2. For python code change, there may also some other python file need update with ACL:https://github.com/search?q=org%3Asonic-net%20redis.Redis&type=code

3. There are code conflict in sonic-swss-common, you may need merge and pass all UT.

**Comment by davidpil2002** [comment]
> In this PR's description, all the code PR link text is "Added support for LDAP". Is it a mistake?

yes - a mistake with the name not with the pointer linked, 
fixed, thanks

File: files/scripts/gen-redis-certs.sh (hunk 2 above)
Comment (davidpil2002) [reply files/scripts/gen-redis-certs.sh:51]: yes - the file was copied to the filesystem with the owner `admin`,
the code can be seen in this PR in the filename
`files/build_templates/sonic_debian_extension.j2`

**Comment by liat-grozovik** [comment]
@Yarden-Z @davidpil2002 are you planning to handle the conflict and to have it ready for merge in 1 week or we should de prioritize it and have it for next release? if needed for 202405 we need you to make it ready and community to prioritize the review.

**Comment by Yarden-Z** [comment]
> @Yarden-Z @davidpil2002 are you planning to handle the conflict and to have it ready for merge in 1 week or we should de prioritize it and have it for next release? if needed for 202405 we need you to make it ready and community to prioritize the review.

We will not make it in time for this release, please move it to the next one.

**Pull request closed without merging**